### PR TITLE
[DNM] Makes NanoMaps Great Again

### PR DESCRIPTION
**Pull request by AffectedArc07**
## What Does This PR Do
This PR overhauls the nanomaps on TGUIs to make them as functional as they were before, if not better. Crew monitor has been used as an example again. A huge thank you to dearmochi here and ShadowLarkens at polaris, as they both wrote major parts of this.
image

image

The zoom is variable from 1x to 8x, in steps of 0.5. The zoom value is also handled fully client side, to reduce overhead of sending to and from the server.

image
Tooltips function as indended while zooming, and the image will zoom uniformally, not from the top left.

AI view
image

## Why It's Good For The Game
My earlier implementation of TGUI nanomaps was added just as a proof of concept, which needed heavy future revisions. This is said future revision, and it brings them on par to what NanoUI has.

## Changelog
:cl: AffectedArc07, dearmochi, ShadowLarkens
add: Overhauled TGUI nanomaps so they dont suck as much. Crew monitor console has been used as an example.
/:cl:

### CODE REVIEW

File: code/modules/tgui/modules/crew_monitor.dm [hunk 1]
```diff
@@ -23,8 +23,8 @@
 /datum/tgui_module/crew_monitor/tgui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = 0, datum/tgui/master_ui = null, datum/tgui_state/state = GLOB.tgui_default_state)
 	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
 	if(!ui)
-		// The 557 may seem random, but its the perfectsize for margins on the nanomap
-		ui = new(user, src, ui_key, "CrewMonitor", name, 1400, 557, master_ui, state)
+		// The 804 may seem random, but its the perfect size for margins on the nanomap
```
Comment (AffectedArc07) [inline code/modules/tgui/modules/crew_monitor.dm:26]: Dont know why I did this comment, the nanomap is a dynamic container now

File: tgui/packages/tgui/interfaces/CrewMonitor.js [hunk 2]
```diff
@@ -1,98 +1,145 @@
...
+            <TableCell>
+              <Box inline
+                color={cm.dead ? 'red' : 'green'}>
+                {cm.dead ? 'Deceased' : 'Living'}
+              </Box>
+              {cm.sensor_type >= 2 ? (
+                <Box inline>
+                  {'('}
+                  <Box inline
+                    color="red">
```
Comment (dearmochi) [inline tgui/packages/tgui/interfaces/CrewMonitor.js:47]: For the sake of consistency, you can probably use the damage type colours defined [here](https://github.com/ParadiseSS13/Paradise/blob/58aa86cb9f9db0fcf9abed19b7c22931f69dfabe/tgui/packages/tgui/constants.js#L21) if they a. 
I don't think they're exported (or defined at all) in CSS which is probably something I would do.

File: tgui/packages/tgui/components/NanoMap.js [hunk 3]
```diff
@@ -1,44 +1,136 @@
...
+      document.removeEventListener('mouseup', this.handleDragEnd);
+    };
+  }
+
+  render() {
+    const { config } = useBackend(this.context);
+    const { offsetX, offsetY } = this.state;
+    const { children, zoom } = this.props;
+
+    const newStyle = {
```
[the remark this answers is not in the document]
Comment (AffectedArc07) [reply tgui/packages/tgui/components/NanoMap.js:74]: This is easier to do than having to modify and get/set/replace existing variables, and I cant think of a decent way to do this

[the remark this answers is not in the document]
Comment (dearmochi) [reply tgui/packages/tgui/components/NanoMap.js:74]: `Object.assign` should do it, otherwise you can probably set them directly by key

**Comment by Carthusia** [comment]
Is it possible with the crew monitor to make it change colors per dot dependent on how wounded they are? Shades of green till it goes red for dead.

**Comment by dearmochi** [comment]
> 
> 
> Is it possible with the crew monitor to make it change colors per dot dependent on how wounded they are? Shades of green till it goes red for dead.

I think it's possible, but that's up to @AffectedArc07 as they made that UI port

**Comment by AffectedArc07** [comment]
Considering we're using fixed colours, and not raw number inputs, I dont think so I also dont know an easy way to make a smooth colour transition which isnt a 20-case switch statmenet.

**Comment by Carthusia** [comment]
curses, even if it was possible to have "Green for minor injuries, orange for severe burns/bruises and red for dead that would do wonders for checking for wounded as a paramedic

**Comment by procdrone** [comment]
I have noticed a few issues I'd love to point out.

With a bigger number of monitored crew, it feels hard to see given there is almost no separation, or its uneven or simply too small. some visual separation between lines would help a ton.

TGUI3 refreshes frequently. Its impossible to find people by using the CTRL+F search function as no results are never shown. Either allowing for searching or adding a dedicated search function would be perfect.

I have also noticed that if you leave the crew monitor open for long time (as for example I do as AI), some entries get corrupted and show up as UNKNOWN and dead but after you close the monitor and reopen it, these entries disappear.

**Comment by Spacemanspark** [comment]
NanoUI also refreshes frequently tbh, would definitely agree with a search feature.

File: code/modules/client/asset_cache.dm [hunk 4]
```diff
@@ -362,3 +361,14 @@ GLOBAL_LIST_EMPTY(asset_datums)
...
+
+// Nanomaps
+/datum/asset/simple/nanomaps
+	// It REALLY doesnt matter too much if these arent up to date
+	// They are friggin huge
+	verify = FALSE
+	assets = list(
+		"Cyberiad_nanomap_z1.png"  = 'icons/_nanomaps/Cyberiad_nanomap_z1.png',
+		"Delta_nanomap_z1.png"  = 'icons/_nanomaps/Delta_nanomap_z1.png',
+		"MetaStation_nanomap_z1.png"  = 'icons/_nanomaps/MetaStation_nanomap_z1.png',
```
Comment (dearmochi) [inline code/modules/client/asset_cache.dm:373]: Align these or remove the double spacing

File: code/modules/tgui/modules/crew_monitor.dm [hunk 5]
```diff
@@ -23,9 +23,13 @@
 /datum/tgui_module/crew_monitor/tgui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = 0, datum/tgui/master_ui = null, datum/tgui_state/state = GLOB.tgui_default_state)
 	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
 	if(!ui)
-		// The 557 may seem random, but its the perfectsize for margins on the nanomap
-		ui = new(user, src, ui_key, "CrewMonitor", name, 1400, 557, master_ui, state)
+		ui = new(user, src, ui_key, "CrewMonitor", name, 800, 600, master_ui, state)
 		ui.autoupdate = TRUE
```
Comment (dearmochi) [inline code/modules/tgui/modules/crew_monitor.dm:27]: Not really needed, [auto-updating](https://github.com/tgstation/tgstation/blob/8a12f15f409e5f2ca9c4c1a561398fe6cdfde09e/code/modules/tgui/tgui.dm#L25) is already on by default

File: code/modules/tgui/modules/crew_monitor.dm [hunk 6]
```diff
@@ -23,9 +23,13 @@
 /datum/tgui_module/crew_monitor/tgui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = 0, datum/tgui/master_ui = null, datum/tgui_state/state = GLOB.tgui_default_state)
```
Comment (dearmochi) [inline code/modules/tgui/modules/crew_monitor.dm:23]: `force_open = FALSE`

File: code/modules/tgui/modules/crew_monitor.dm [hunk 7]
```diff
@@ -23,9 +23,13 @@
...
 	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
 	if(!ui)
-		// The 557 may seem random, but its the perfectsize for margins on the nanomap
-		ui = new(user, src, ui_key, "CrewMonitor", name, 1400, 557, master_ui, state)
+		ui = new(user, src, ui_key, "CrewMonitor", name, 800, 600, master_ui, state)
 		ui.autoupdate = TRUE
+
+		// Send nanomaps
+		var/datum/asset/nanomaps = get_asset_datum(/datum/asset/simple/nanomaps)
+		nanomaps.send(user)
```
Comment (dearmochi) [inline code/modules/tgui/modules/crew_monitor.dm:31]: I'd personally prefer if assets were sent *before* the UI is opened

File: tgui/packages/tgui/interfaces/CrewMonitor.js [hunk 8]
```diff
@@ -1,98 +1,148 @@
...
+          </Table.Cell>
+          <Table.Cell>
+            Status
+          </Table.Cell>
+          <Table.Cell>
+            Location
+          </Table.Cell>
+        </Table.Row>
+        {crew.map(cm => (
+          <Table.Row key={cm.name}>
```
Comment (dearmochi) [inline tgui/packages/tgui/interfaces/CrewMonitor.js:35]: For in-game readability, I think lines should be spaced by a small margin (`0.25rem` or `0.5rem`) and be highlighted on hover. You can use the `onMouseEnter`/`onMouseLeave` events OR make a dedicated `.scss` file (which I would prefer!)

File: tgui/packages/tgui/interfaces/CrewMonitor.js [hunk 9]
```diff
@@ -1,98 +1,148 @@
...
+            {crew.filter(x => x.sensor_type === 3).map(cm => (
+              <NanoMap.Marker
+                key={cm.ref}
+                x={cm.x}
+                y={cm.y}
+                zoom={mapZoom}
+                icon="circle"
+                tooltip={cm.name}
+                color={cm.dead ? 'red' : 'green'}
+              />
```
Comment (dearmochi) [inline tgui/packages/tgui/interfaces/CrewMonitor.js:115]: When I tested locally, markers were visible at zoom level 1x then disappeared if I increased it by even 0.1.

File: code/modules/tgui/modules/crew_monitor.dm (hunk 7 above)
Comment (AffectedArc07) [reply code/modules/tgui/modules/crew_monitor.dm:31]: `ui.open` is on line 33, this is on line 31.

File: tgui/packages/tgui/interfaces/CrewMonitor.js (hunk 9 above)
Comment (AffectedArc07) [reply tgui/packages/tgui/interfaces/CrewMonitor.js:115]: Cannot reproduce

**Comment by AffectedArc07** [comment]
Someone else can pick this up if they want. I keep running into issues with assets and scaling and pain.

**Pull request closed without merging**